### DISCUSSION TITLE
Building custom DB Adapter and custom DB driver

### DISCUSSION THREAD

**Discussion by ABSanthosh** [prisma/prisma#24191] [Q&A]
### Question

Here's my usecase:

Assuming the data store is a csv file:-
- `prisma db push`  
  - will make the columns of the csv file
  - Generate types 
  - Generate Client functions like the following for each of the column in node_modules>.prisma>client>index.d.ts
```
    /**
     * Create a Cart.
     * @param {CartCreateArgs} args - Arguments to create a Cart.
     * @example
     * // Create one Cart
     * const Cart = await prisma.cart.create({
     *   data: {
     *     // ... data to create a Cart
     *   }
     * })
     * 
    **/
    create<T extends CartCreateArgs<ExtArgs>>(
      args: SelectSubset<T, CartCreateArgs<ExtArgs>>
    ): Prisma__CartClient<$Result.GetResult<Prisma.$CartPayload<ExtArgs>, T, 'create'>, never, ExtArgs>
```
- Then user can use `prisma.cart.create({...})` to add a new item in the csv
- In the future, when the user finally no longer want to use the csv, they can just swap out "csv" from the `provider` in the `datasource` block and not touch the `prisma.cart.xyz({...})` throughout the app.

![image](https://github.com/nextauthjs/next-auth/assets/24393343/fd1f5e22-80b8-421c-ae54-8258303df127)

Is this possible?

**Comment by ludralph** [comment]
Hi @ABSanthosh 👋 

Thank you for raising this question.

Using a CSV file as a data source is not natively supported in Prisma. Prisma is designed to work with databases like PostgreSQL, MySQL, SQLite, SQL Server, and MongoDB.

The functionality you described, being able to use `prisma.cart.create({...})` to add items to a CSV file and later seamlessly switch the provider to a relational database without changing application code—requires a level of abstraction and flexibility that Prisma does not natively provide for CSV files.

You would have to write some custom logic to read and parse the CSV file before writing it using Prisma and storing in any of the supported databases by Prisma ORM.

You can also fill out a [feature request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) detailing your usecase.

**Reply by ABSanthosh** [reply]
Hi @ludralph!

Thanks for taking your time to answer my questions. 

> requires a level of abstraction and flexibility that Prisma does not natively provide for CSV files.

I'm quite aware of this and I'm looking to build my own abstraction but I can't seem to find any proper documentation for the same. 

From what I discussed with @jharrell over on discord, I believe I'll have to use the SQL that comes out of the query engine and process it using my custom database driver to interact with csv.
![image](https://github.com/prisma/prisma/assets/24393343/d1c10cbe-7eb0-4f5a-86aa-81d8b7a27934)

I'm looking for
- Documentation for writing a database driver
- Documentation for setting up the dev environment
- Documentation for bundling it into a library to upload to NPM

I hope you can help me with these. 

Edit: I found a community made db-driver and I'm tying to understand how it works. It seems to use a helper library called [@prisma/driver-adapter-utils](https://www.npmjs.com/package/@prisma/driver-adapter-utils) but there are no documentation for it.

If you can just share how to setup the dev env, I'll figure the rest out by myself(hopefully)

**Reply by ceddybi** [reply]
@ABSanthosh how far?

**Reply by ABSanthosh** [reply]
@ceddybi Sorry?

**Reply by ceddybi** [reply]
> @ceddybi Sorry?

@ABSanthosh 
were you able to create a cusom adapter for your database?, i'm looking to create one for couchbase

**Reply by ABSanthosh** [reply]
@ceddybi I haven't gotten around figuring out the dev set up. From what I know, most of the documentation for these tools are internal and if you want to do anything, you'd have to read the code and figure it out ourselves. Hope @ludralph gets back with some documentation soon.

**Reply by anta-semenov** [reply]
[This](https://github.com/prisma/prisma/blob/main/packages/driver-adapter-utils/src/types.ts) could be a good starting point to see what Prisma expects from the adapter.

Basically it should have three methods
```
startTransaction()
queryRaw()
executeRaw()
```

And two attributes
```
provider // as I understand only three are supported right now and define what SQL dialect prisma will use in raw sql
adapterName // could be a string
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
